Mathlib/CategoryTheory/MorphismProperty/Lifting.lean

```diff
@@ -0,0 +1,159 @@
+/-
+Copyright (c) 2024 Jack McKoen. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jack McKoen
+-/
+import Mathlib.CategoryTheory.MorphismProperty.Basic
+import Mathlib.CategoryTheory.MorphismProperty.Limits
+
+/-!
+# Left and right lifting properties
+
+Given a morphism property `T`, we define the left and right lifting property with respect to `T`.
+
+We show that the left lifting property is stable under retracts, cobase change, and composition,
+with dual statements for the right lifting property.
+
+-/
+
+universe v u
+
+namespace CategoryTheory
+
+variable {C : Type u} [Category.{v} C]
+
+/-- `f : X ⟶ Y` is a retract of `g : Z ⟶ W` if there are morphisms `i : f ⟶ g`
+and `r : g ⟶ f` in the arrow category of `C` such that `i ≫ r = 𝟙 f`. -/
+class IsRetract {X Y Z W : C} (f : X ⟶ Y) (g : Z ⟶ W) where
+  /-- `i : f ⟶ g` -/
+  i : Arrow.mk f ⟶ Arrow.mk g
+  /-- `r : g ⟶ f` -/
+  r : Arrow.mk g ⟶ Arrow.mk f
+  /-- `i ≫ r = 𝟙 f` -/
+  retract : i ≫ r = 𝟙 Arrow.mk f
+
+namespace MorphismProperty
+
+section Retract
+
+/-- A class of morphisms is stable under retracts if the retract of a morphism still
+lies in the class. -/
+class IsStableUnderRetracts (P : MorphismProperty C) : Prop where
+  of_isRetract {X Y Z W : C} {f : X ⟶ Y} {g : Z ⟶ W} (h : IsRetract f g) (hg : P g) : P f
+
+lemma of_isRetract {P : MorphismProperty C} [P.IsStableUnderRetracts]
+    {X Y Z W : C} {f : X ⟶ Y} {g : Z ⟶ W} (h : IsRetract f g) (hg : P g) : P f :=
+  IsStableUnderRetracts.of_isRetract h hg
+
+instance IsStableUnderRetracts.monomorphisms :
+    (monomorphisms C).IsStableUnderRetracts := by
+  refine ⟨fun {X Y} _ _ f g H p ↦ ⟨fun α β ω ↦ ?_⟩⟩
+  have h : H.i.left ≫ g = f ≫ H.i.right := H.i.w
+  have := ω =≫ H.i.right
+  rw [Category.assoc, Category.assoc, ← h, ← Category.assoc, ← Category.assoc] at this
+  have ω' := p.right_cancellation (α ≫ H.i.left) (β ≫ H.i.left) this =≫ H.r.left
+  have : H.i.left ≫ H.r.left = 𝟙 X := Arrow.hom.congr_left H.retract
+  rwa [Category.assoc, Category.assoc, this, Category.comp_id, Category.comp_id] at ω'
+
+end Retract
+
+/-- The left lifting property (llp) with respect to a class of morphisms. -/
+def Llp (T : MorphismProperty C) : MorphismProperty C := fun _ _ f ↦
+  ∀ ⦃X Y : C⦄ ⦃g : X ⟶ Y⦄ (_ : T g), HasLiftingProperty f g
+
+/-- The right lifting property (rlp) with respect to a class of morphisms. -/
+def Rlp (T : MorphismProperty C) : MorphismProperty C := fun _ _ f ↦
+  ∀ ⦃X Y : C⦄ ⦃g : X ⟶ Y⦄ (_ : T g), HasLiftingProperty g f
+
+lemma Llp.IsStableUnderRetracts {T : MorphismProperty C} : T.Llp.IsStableUnderRetracts := by
+  refine ⟨fun {X Y} _ _ f f' H L _ _ g h ↦ ⟨fun {u v} sq ↦ ?_⟩⟩
+  have : (H.r.left ≫ u) ≫ g = f' ≫ (H.r.right ≫ v) := by simp [sq.w]
+  obtain lift := ((L h).sq_hasLift (CommSq.mk this)).exists_lift.some
```

Comment:
The `HasLiftingProperty` API should be improved in order to provide a choice of a lifting and both `fac_left/fac_right` lemmas. This could defined as `CommSq.lift` under a suitable `[HasLiftingProperty i p]` assumption.